From 1ecdf1c8abf76db5863b1cc71fd8a77a6cc28cd2 Mon Sep 17 00:00:00 2001
From: Pratham Patel <ppatel@ciq.com>
Date: Mon, 11 Aug 2025 21:44:47 +0530
Subject: [PATCH] userfaultfd: fix a race between writeprotect and exit_mmap()

jira VULN-4366
cve CVE-2021-47461
commit-author Nadav Amit <namit@vmware.com>
commit cb185d5f1ebf900f4ae3bf84cee212e6dd035aca

A race is possible when a process exits, its VMAs are removed by
exit_mmap() and at the same time userfaultfd_writeprotect() is called.

The race was detected by KASAN on a development kernel, but it appears
to be possible on vanilla kernels as well.

Use mmget_not_zero() to prevent the race as done in other userfaultfd
operations.

Link: https://lkml.kernel.org/r/20210921200247.25749-1-namit@vmware.com
Fixes: 63b2d4174c4ad ("userfaultfd: wp: add the writeprotect API to userfaultfd ioctl")
	Signed-off-by: Nadav Amit <namit@vmware.com>
	Tested-by: Li  Wang <liwang@redhat.com>
	Reviewed-by: Peter Xu <peterx@redhat.com>
	Cc: Andrea Arcangeli <aarcange@redhat.com>
	Cc: <stable@vger.kernel.org>
	Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
	Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
(cherry picked from commit cb185d5f1ebf900f4ae3bf84cee212e6dd035aca)
	Signed-off-by: Pratham Patel <ppatel@ciq.com>
---
 fs/userfaultfd.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/fs/userfaultfd.c b/fs/userfaultfd.c
index 9ba1fa459921..7a6f50555952 100644
--- a/fs/userfaultfd.c
+++ b/fs/userfaultfd.c
@@ -1801,9 +1801,15 @@ static int userfaultfd_writeprotect(struct userfaultfd_ctx *ctx,
 	if (mode_wp && mode_dontwake)
 		return -EINVAL;
 
-	ret = mwriteprotect_range(ctx->mm, uffdio_wp.range.start,
-				  uffdio_wp.range.len, mode_wp,
-				  &ctx->mmap_changing);
+	if (mmget_not_zero(ctx->mm)) {
+		ret = mwriteprotect_range(ctx->mm, uffdio_wp.range.start,
+					  uffdio_wp.range.len, mode_wp,
+					  &ctx->mmap_changing);
+		mmput(ctx->mm);
+	} else {
+		return -ESRCH;
+	}
+
 	if (ret)
 		return ret;
 
-- 
2.47.3

